extensions/objects/templates/audioFileEditor.js

```diff
@@ -0,0 +1,128 @@
+      ContextService, $timeout) {
+    return {
+      restrict: 'E',
+      scope: {
+        getAlwaysEditable: '&',
+        value: '='
+      },
+      templateUrl: UrlInterpolationService.getExtensionResourceUrl(
+        '/objects/templates/audio_file_editor_directive.html'),
+      controller: ['$scope', function($scope) {
+        $scope.alwaysEditable = $scope.getAlwaysEditable();
+        var OUTPUT_AUDIO_FORMAT = '.mp3';
+        var FILE_ELEMENT = 'audio_file';
+        var URL_ELEMENT = 'audio_url';
+
+        var generateAudioFilename = function() {
+          var date = new Date();
+          return 'audio_' +
+            date.getFullYear() +
+            ('0' + (date.getMonth() + 1)).slice(-2) +
+            ('0' + date.getDate()).slice(-2) +
+            '_' +
+            ('0' + date.getHours()).slice(-2) +
+            ('0' + date.getMinutes()).slice(-2) +
+            ('0' + date.getSeconds()).slice(-2) +
+            '_' +
+            Math.random().toString(36).substr(2, 10) +
+            OUTPUT_AUDIO_FORMAT;
+        };
+        var validateData = function() {
+          audioFile = fileElement.files[0];
+          audioUrl = urlElement.value;
+          if (audioFile && audioUrl) {
+            return true;
+          } else if (audioFile) {
+            return true;
+          } else if (audioUrl) {
+            return true;
+          } else {
+            return false;
+          }
+        };
+
+        var downloadAndSaveAudio = function() {
+          $scope.warningText = '';
+          if (validateData()) {
+            $scope.data.showProgress = true;
+            var filename = generateAudioFilename();
+            var data = {};
+            if (audioFile) {
+              data.audioFile = audioFile;
+            } else {
+              data.audioUrl = audioUrl;
+            }
+            AssetsBackendApiService.downloadAndSaveAudio(
+              ContextService.getExplorationId(),
+              filename, data
+            ).then(function() {
+              var downloadUrl = AssetsBackendApiService.getAudioDownloadUrl(
+                ContextService.getExplorationId(), filename);
+              $scope.data.showProgress = false;
+              $scope.data.downloadUrl = downloadUrl;
+              $scope.value = downloadUrl;
+              $timeout(function() {
+                $scope.$apply();
+              }, 0, false);
+            }, function(errorResponse) {
+              $scope.warningText = errorResponse.error;
+              $scope.data.downloadUrl = null;
+              $scope.data.showProgress = false;
+              $timeout(function() {
+                $scope.$apply();
+              }, 0, false);
+            });
+          }
+        };
+
+        $scope.uploadAudioFromUrl = function() {
+          audioFile = null;
+          downloadAndSaveAudio();
+        };
+        $scope.validate = function(data) {
+          // return !data.showProgress && data.downloadUrl;
```

Comment:
is this expected?

Reply:
Hmm, in that case I suggest filing a bug (after this PR is merged) pointing to this part of the code and describing how we need to disable the done button when there is no audio.